astropy/table/table.py

```diff
@@ -2867,30 +2911,53 @@ def __ge__(self, other):
 
     def __eq__(self, other):
 
-        if isinstance(other, Table):
-            other = other.as_array()
+        # Check the length of other
+        check_len = True
 
-        if self.has_masked_columns:
-            if isinstance(other, np.ma.MaskedArray):
-                result = self.as_array() == other
-            else:
-                # If mask is True, then by definition the row doesn't match
-                # because the other array is not masked.
-                false_mask = np.zeros(1, dtype=[(n, bool) for n in self.dtype.names])
-                result = (self.as_array().data == other) & (self.mask == false_mask)
+        if isinstance(other, Table):
+            names = other.colnames
         else:
-            if isinstance(other, np.ma.MaskedArray):
-                # If mask is True, then by definition the row doesn't match
-                # because the other array is not masked.
-                false_mask = np.zeros(1, dtype=[(n, bool) for n in other.dtype.names])
-                result = (self.as_array() == other.data) & (other.mask == false_mask)
+            try:
+                other = Table(other, copy=False)
+                names = other.colnames
+            except Exception:
+                # Broadcast other into a dict, so e.g. other = 2 will turn into
+                # other = {'a': 2, 'b': 2} and then equality does a 
+                # column-by-column broadcasting.
+                names = self.colnames
+                other = {name: other for name in names}
+                check_len = False
+
+        if check_len and len(self) != len(other):
+            raise ValueError('cannot compare tables with different lengths')
+
+        # Require column names match but do not require same column order
+        if set(self.colnames) != set(names):
+            raise ValueError('cannot compare tables with different column names')
+
+        eqs = []
+        for name in names:
+            try:
+                eq = self[name] == other[name]
+                assert isinstance(eq, np.ndarray) and eq.dtype is np.dtype('bool')
+            except Exception:
+                raise ValueError(f'unable to compare column {name}')
             else:
-                result = self.as_array() == other
+                if np.any([str(x) == '--' for x in list(eq)]) != True:
```

Comment:
Except for when elements are masked converting to list because 
when, 
```
t1 = Table(masked=True)
t1['c'] = ['c','a','b']
t = simple_table()
t['c'] == t1['c']
Out[9]: 
masked_array(data=[ True, False, False],
             mask=False,
       fill_value='N/A')
```
Which cannot be inserted in a table.
So, first converting the output to list and then storing in `out`.

Reply:
In astropy table code, you have to worry about performance and memory.  Any time you find yourself turning something into a pure Python list e.g. `list(eq)` or doing element by element operations, then that is a red flag and it is very possible the reviewer will suggest a change.  Remember that `eq` could be 1e9 elements long!

See if you can figure out a different way, otherwise I'll think about it.

Reply:
I changed that behavior please have a look. It is an irony that the person who wants to increase `table` performance not taking care of performance in his code I will take care of it next time. Sorry for irresponsible behavior :sweat:

Reply:
It's all a learning process!